This thesis explores the prediction of protein abundance from mRNA expression data in cancer cell lines using a combination of linear and deep learning models. The study leverages data from the Cancer Cell Line Encyclopedia (CCLE), incorporating both transcriptomic and proteomic profiles to understand the complex relationships between gene expression and protein levels. The research begins with the application of linear models to identify proteins that can be predicted accurately using straightforward methods, emphasizing the balance between simplicity and computational efficiency. Subsequently, more sophisticated approaches, including autoencoders and the CarNet model (which integrates convolutional layers, LSTM layers, and self-attention mechanisms), are employed to capture non-linear relationships and enhance prediction accuracy. The findings demonstrate that while deep learning models show promise, their performance is often comparable to simpler linear models, highlighting the potential of hybrid approaches. The study underscores the importance of feature selection, using PCA and ANOVA to identify the most influential genes, and gives valuable insights on Pearson and Spearman correlations and their use to predict proteins with low degree of further information
